### DISCUSSION TITLE
Smtp server and dkim

### DISCUSSION THREAD

**Discussion by mjac1521** [stalwartlabs/smtp-server#5] [General]
Hi guys super incredible how easy, simplified and clean all the instalation of smtp, jmap and imap servers are. Well done.

I habe a question. I installed jmap and created multiple domains and each of it has it own dkim generated using the jmap cli and stored in rockdb. I have also installed the smtp server wich is autenticating through imap wich get the user info from jmap I also setup lmtp to send incoming emails to jmap. 

My question is if an email client let say outlook send an email from any domain to an external email through the smtp server will the smtp server be able to get the dkim key for the specific domain stored in the jmap rockdb and attached to the email?

Sincerely Maykel

**Comment by mdecimus** [comment]
Hi Maykel,

At the moment Stalwart SMTP and JMAP are not fully integrated and each one uses their own DKIM keys. So, if you send an email from Outlook the message will be signed using the DKIM key configured in the SMTP server. And if you send a message using JMAP submissions, it will be signed using the keys configured in RocksDB.

However this is for the time being, both servers are currently being integrated and the next release will allow DKIM keys to be configured only in a single place.

**Reply by Uganto** [reply]
who knows how to build an SMTP?

**Comment by mjac1521** [comment]
Thank you for the fast answer. One last question, is it possible for the smtp server to handler dkim for multiple domain? I mean could I store the dkim information like key, cert, selector, domain, etc. in mysql and have the smtp server get it from the database? 

Sincerely Maykel

**Reply by mdecimus** [reply]
It is possible to enable DKIM for multiple domains but the settings/certificates cannot be read from mySQL at the moment unfortunately.

**Reply by smorks** [reply]
hello! i'm trying to get DKIM signing working with multiple domains.

i've managed to get it working with a single domain, using the following config:

```toml
sign = [ { if = "listener", ne = "smtp", then = ["domain1"] },
         { else = [] } ]
```

but i'm not super clear on how to handle multiple. i tried the following:

```toml
[auth.dkim]
sign = [ { all-of = [ { if = "listener", ne = "smtp" },
                      { if = "sender-domain", eq = "domain1.com" } ], then = ["domain1"] },
       { all-of = [ { if = "listener", ne = "smtp" },
                    { if = "sender-domain", eq = "domain2.net" } ], then = ["domain2"] },
       { else = [] } ]
```

but then the e-mails aren't getting signed at all. is there something i'm doing wrong or is there a better way to do it?

**Reply by mdecimus** [reply]
Are you sending the message using a listener different than `smtp`? In other words, are you sending the message using the submissions port? If you are, then you can try the following:

- Enable debug or tracing log level to see if there are any problems during signing.
- If no errors are recorded, then test the signatures individually without any conditions:

```
[auth.dkim]
sign = "domain1"
```

and then:

```
[auth.dkim]
sign = "domain2"
```

**Reply by smorks** [reply]
yes, i'm using the `submissions` listener to send messages.

i tried enabling tracing, but there were no errors recorded.

i've also tried both signatures individually, like you suggested, and together `sign = ["domain1", "domain2"]`, and it all works as expected.

i will also try without the `listener` part, to see if that makes any difference. something like this:

```toml
sign = [ { if = "sender-domain", eq = "domain1.com", then = ["domain1"] },
       { if = "sender-domain", eq = "domain2.net", then = ["domain2"] },
       { else = [] } ]
```

and will report back.

**Reply by smorks** [reply]
for simplicity's sake, i just tried it with one domain, as follows:

```toml
sign = [ { if = "sender-domain", eq = "domain1.com", then = ["domain1"] },
       { else = [] } ]
```

and it didn't work. so my guess would be the `if = "sender-domain"` isn't working.

i'll take a look through the code to see if anything sticks out to me.

**Reply by mdecimus** [reply]
Thanks, I'll give it a look tomorrow as well.

**Reply by mdecimus** [reply]
I've just released version `0.1.2` which should fix this problem.

**Reply by smorks** [reply]
awesome, thanks!

**Reply by mjac1521** [reply]
Hi, 

In order to upgrade I just ran the same command used to install
curl --proto '=https' --tlsv1.2 -sSf https://smtp.stalw.art/install.sh | sudo sh

From the CLI
root:/home/ubuntu# /usr/local/stalwart-smtp/bin/stalwart-cli --version
stalwart-cli 0.2.0

From the smtp log
Starting Stalwart SMTP server v0.1.1..

From your last comment it should upgrade to 0.1.2. I am not sure if it was really upgraded or not.

Sincerely Maykel

**Reply by mdecimus** [reply]
I haven't had the chance to create an upgrade script yet. But since this is a minor upgrade you can simply download the new binary from https://github.com/stalwartlabs/smtp-server/releases/tag/v0.1.2 and replace it under `/usr/local/stalwart-smtp/bin`.

**Reply by smorks** [reply]
> I've just released version `0.1.2` which should fix this problem.

it's working as expected now, thanks again!
